pkg/controller/ingressstate/ingress_state_controller.go

```diff
@@ -0,0 +1,328 @@
+func loadServiceCA() (*x509.CertPool, error) {
+	bundlePath := "/var/run/configmaps/service-ca-bundle/service-ca.crt"
+	bundlePEM, err := ioutil.ReadFile(bundlePath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading service ca bundle: %v", err)
+	}
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(bundlePEM)
+	if !ok {
+		return nil, fmt.Errorf("no certificates could be parsed from the service ca bundle")
+	}
+	return roots, nil
+}
+
+// checkEndpointHealthz will check the health of given https://endpointIP:6443/healthz
+func checkEndpointHealthz(endpointIP string, rootCAs *x509.CertPool) error {
+	client := http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: rootCAs,
+			},
+		},
+	}
+	url := fmt.Sprintf("https://%s:6443/healthz", endpointIP)
+	response, err := client.Get(url)
+	if err != nil {
+		reportedError := err
+		if response != nil {
+			reportedError = fmt.Errorf("status:%q, body: %q, error: %v", response.Status, response.Body, reportedError)
+		}
+		return reportedError
+	}
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("status:%q, body: %q", response.Status, response.Body)
+	}
+	return nil
+}
+
+// Providing these helper functions as arguments to ingressConditions
+// supports substituting them in testing.
+type checkPodFunc func(reference *corev1.ObjectReference) []string
+type checkEndpointsFunc func(endpointIP string) error
+
+// degradedIngressConditions checks the health of a set of endpoints
+// and returns degraded operator conditions if any of the endpoints or
+// the pods that back them are unhealthy.
+func degradedIngressConditions(endpoints *corev1.Endpoints, checkPod checkPodFunc, checkEndpoints checkEndpointsFunc) ([]operatorv1.OperatorCondition, error) {
+	if endpoints == nil {
+		return []operatorv1.OperatorCondition{
+			endpointsDegraded("MissingEndpoints", "No endpoints found for oauth-server"),
+		}, nil
+	}
+
+	if len(endpoints.Subsets) == 0 {
+		return []operatorv1.OperatorCondition{
+			endpointsDegraded("ZeroEndpointsSubsets", "No endpoints found for oauth-server"),
+		}, nil
+	}
+
+	conditions := []operatorv1.OperatorCondition{}
+
+	// Aggregate unhealthy endpoints and pods across all subsets to
+	// support going degraded only if more than one endpoint or pod is
+	// unhealthy. A single endpoint or pod would commonly be unhealthy
+	// during an upgrade, and it is not desirable to report degraded
+	// in that case.
+	unhealthyEndpoints := []string{}
+	unhealthyPodCount := 0
+	unhealthyPodMessages := []string{}
+
+	for _, subset := range endpoints.Subsets {
+		// Report degraded if all addresses for the endpoint are not ready
+		if len(subset.Addresses) == 0 && len(subset.NotReadyAddresses) > 0 {
+			msg := fmt.Sprintf("%d endpoints for oauth-server are reporting 'not ready'", len(subset.NotReadyAddresses))
+			conditions = append(conditions, endpointsDegraded("NonReadyEndpoints", msg))
```

Comment:
this (and later uses of the `endpointsDegraded` helper) overrides the `IngressStateEndpointsDegraded` condition with the latest msg, is that intended?

Reply:
Hah! No. 

I find it easy to be ignorant when I inherit code in an unfamiliar domain (e.g. endpoints). I've updated the code with an assumption that I think makes the code simpler to reason about - that there will only ever be a single subset. This is true so long as the service exposes only a single port.